internal/elfexec/elfexec.go

```diff
@@ -240,17 +240,10 @@ func GetBase(fh *elf.FileHeader, loadSegment *elf.ProgHeader, stextOffset *uint6
 		}
 		return start, nil
 	case elf.ET_DYN:
-		if offset != 0 {
-			if loadSegment == nil || loadSegment.Vaddr == 0 {
-				return start - offset, nil
-			}
-			return 0, fmt.Errorf("Don't know how to handle mapping. Offset=%x, vaddr=%x",
-				offset, loadSegment.Vaddr)
-		}
 		if loadSegment == nil {
-			return start, nil
+			return start - offset, nil
 		}
-		return start - loadSegment.Vaddr, nil
+		return start - offset + loadSegment.Off - loadSegment.Vaddr, nil
```

Comment:
The logic looks correct to me.  I agree that loadSegment.Off should be accounted.  Basically, (start - offset) is the virtual address of where the file offset of zero is (as if the file is mapped into memory as a whole).  Then we add the offset of the segment from the program header which gives the virtual address of where the segment starts.  Then we subtract the virtual address of where the program headers the segment should start which gives the adjustment that we return.

In practice, the executable segment of shared libraries has typically file offset of zero I think which is why we didn't notice it until we started dealing with non-standard ELF files.

Can you put all of this reasoning into the code as a comment?  I think it's subtle enough and deserves more than just being a PR discussion.